boskos/ranch/priority.go

```diff
@@ -0,0 +1,219 @@
+func (rq *requestQueue) cleanup(now time.Time) {
+	rq.lock.Lock()
+	defer rq.lock.Unlock()
+	var newRequestList []string
+	newRequestMap := map[string]request{}
+	for _, requestID := range rq.requestList {
+		req := rq.requestMap[requestID]
+		// Checking expiration
+		if now.After(req.expiration) {
+			logrus.Infof("request id %s expired", req.id)
+			continue
+		}
+		// Keeping
+		newRequestList = append(newRequestList, requestID)
+		newRequestMap[requestID] = req
+	}
+	rq.requestMap = newRequestMap
+	rq.requestList = newRequestList
+}
+
+// getRank provides the rank of a given requestID following the order it was added (FIFO).
+// If requestID is an empty string, getRank assumes it is added last (lowest rank + 1).
+func (rq *requestQueue) getRank(requestID string, ttl time.Duration, now time.Time) int {
+	rq.update(requestID, now.Add(ttl))
+	rank := 1
+	rq.lock.RLock()
+	defer rq.lock.RUnlock()
+	for _, existingID := range rq.requestList {
+		req := rq.requestMap[existingID]
+		if now.After(req.expiration) {
+			logrus.Infof("request id %s expired", req.id)
+			continue
+		}
+		if requestID == existingID {
+			return rank
+		}
+		rank++
+	}
+	return rank
+}
+
+func (rq *requestQueue) isEmpty() bool {
+	rq.lock.Lock()
+	defer rq.lock.Unlock()
+	return len(rq.requestList) == 0
+}
+
+// RequestManager facilitates management of RequestQueues for a set of (resource type, resource state) tuple.
+type RequestManager struct {
+	lock     sync.Mutex
+	requests map[interface{}]*requestQueue
+	ttl      time.Duration
+	stopGC   context.CancelFunc
+	wg       sync.WaitGroup
+	// For testing only
+	now func() time.Time
+}
+
+// NewRequestManager creates a new RequestManager
+func NewRequestManager(ttl time.Duration) *RequestManager {
+	return &RequestManager{
+		requests: map[interface{}]*requestQueue{},
+		ttl:      ttl,
+		now:      time.Now,
+	}
+}
+
+func (rp *RequestManager) cleanup(now time.Time) {
+	rp.lock.Lock()
+	defer rp.lock.Unlock()
+	for key, rq := range rp.requests {
+		logrus.Infof("cleaning up %v request queue", key)
+		rq.cleanup(now)
+		if rq.isEmpty() {
+			delete(rp.requests, key)
+		}
+	}
+}
+
+// StartGC starts a goroutine that will call cleanup every gcInterval
+func (rp *RequestManager) StartGC(gcPeriod time.Duration) {
+	ctx, stop := context.WithCancel(context.Background())
+	rp.stopGC = stop
+	tick := time.Tick(gcPeriod)
+	go func() {
+		logrus.Info("starting cleanup go routine")
+		defer logrus.Info("exiting cleanup go routine")
+		rp.wg.Add(1)
+		defer rp.wg.Done()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-tick:
+				rp.cleanup(rp.now())
+			}
+		}
+
+	}()
+}
+
+// StopGC is a blocking call that will stop the GC goroutine.
+func (rp *RequestManager) StopGC() {
+	if rp.stopGC != nil {
+		rp.stopGC()
+		rp.wg.Wait()
+	}
+}
+
+// GetRank provides the rank of a given request
+func (rp *RequestManager) GetRank(key interface{}, id string) int {
```

Comment:
Why does the `key` need to be `interface{}`?

Reply:
I was thinking that this is only used in Acquire, but could be used elsewhere, and you might want to define your own struct as the key.

Reply:
In the future we should handle other cases but for now we should just make it a struct

Reply:
Using an interface now does not require extra work to be used, so I don't think this is a valid request.